docs/content/sips/draft-000-templates.md

```diff
@@ -0,0 +1,365 @@
+title = "SIP DRAFT-000 - Templates"
+template = "main"
+date = "2022-03-22T14:53:30Z"
+
+---
+
+Summary: Proposes changes to the `spin new` syntax and to the template
+authoring format
+
+Created: March 28, 2022
+
+Status: DRAFT
+
+## Background
+
+We have an initial cut of a template system planned for 0.1, but design and UX
+work on this was somewhat constrained by the 0.1 deadline.  The 0.1 design is
+closely based on `cargo generate` because it was a readily available, er,
+template in the language we were using.  This proposal discusses goals and
+design options for taking the template system forward.
+
+The existing work gives us a solid basis with a well-documented template syntax
+using the Liquid engine, safe failure modes via staging, and other good
+stuff.  No changes are proposed to that core.  This SIP relates to user
+experience (for both consumers and authors) and broader architecture.
+
+## Goals
+
+* Convenient, forgiving user experience for consumers
+* Foundation for delivery via IDE
+* Minimise what authors need to know
+
+## Should we have a template system at all?
+
+It's worth considering this because there are a ton of existing templating
+systems, some language specific (e.g. `cargo generate`, `dotnet new`), some
+generic (e.g. Yeoman).  These are likely to be idiomatic to the languages
+people use.  Is it worth having our own?
+
+Reasons why we should:
+
+* Some languages _don't_ have template support, or require the user to go
+  install yet more tools.
+* It allows documentation to assume Spin templates instead of having to
+  write different docs for every language.
+* Having our own system doesn't preclude offering Spin templates via
+  language-idiomatic systems too.
+
+Reasons why we shouldn't:
+
+* It's a decent chunk of work to do it well.
+* It's extra code for every copy of Spin to carry around.
+
+I believe the general consensus is that the case for having templates built
+in is compelling, but it's always worth checking that assumption.
+
+## Identifying a template
+
+Templates are currently identified by a repository ID and directory name.
+Because Spin is a multi-language runtime, it would be good to also identify
+templates by language, so that a user could easily see and choose templates
+in their preferred language.
+
+The question of the repository ID is what we want "adding" a template to mean.
+Does adding a template:
+
+* Record a mapping of a repository ID to a Git location
+* Copy a template to the local system
```

Comment:
It seems to me that this should be handled by caching (for remote git repos).

Reply:
You mean cloning the entire repo into the local environment?  That's what we do at the moment; it feels heavy to me but it's absolutely a viable way.

Reply:
I just mean _if_ there is some compelling (performance?) reason to have a local copy, treat it as a cache rather than the primary result of "adding" a template.

Reply:
I'm not too concerned about performance, though the offline aspect bears thinking about (ye olde "working on my conference demos on the plane to the conference" scenario).  What I'm concerned about is convenience - being able to type `spin new http` rather than bashing out GitHub URLs - and discoverability - being able to browse the templates one has "installed" or "favourited" or however we choose to think of it, rather than having the whole of URL-space to choose from _grin_.  I should probably be clearer in tying the usage experience back to the "goals" section.